rllib/policy/sample_batch.py

```diff
@@ -133,27 +134,68 @@ def __init__(self, *args, **kwargs):
         if self._is_training is None:
             self._is_training = self.pop("is_training", False)
 
+        # Try to infer the "length" of the SampleBatch by finding the first
+        # value that is actually a ndarray/tensor.
         lengths = []
         copy_ = {k: v for k, v in self.items() if k != SampleBatch.SEQ_LENS}
         for k, v in copy_.items():
             assert isinstance(k, str), self
 
-            # TODO: Drop support for lists as values.
+            # TODO: Drop support for lists and Numbers as values.
             # Convert lists of int|float into numpy arrays make sure all data
             # has same length.
-            if isinstance(v, list):
+            if isinstance(v, (Number, list)):
                 self[k] = np.array(v)
 
-            # Try to infer the "length" of the SampleBatch by finding the first
-            # value that is actually a ndarray/tensor. This would fail if
-            # all values are nested dicts/tuples of more complex underlying
-            # structures.
+            if k == SampleBatch.INFOS:
+                # Don't attempt to count on infos since we make no assumptions
+                # about its content
+                continue
+
+            # If this is a nested dict (for example a nested observation),
+            # try to flatten it, assert that all elements have the same length (batch
+            # dimension)
+            v_list = tree.flatten(v) if isinstance(v, (dict, tuple)) else [v]
```

Comment:
maybe you don't need this if ... else. tree.flatten specifically converts single obj into [obj]
```
>>> tree.flatten(np.array([3, 4]))
[array([3, 4])]
```

Reply:
Which else are you referring to? Because the upper one belongs to the try/except clause, not the if.

This tree flattening also converts nested lists and therefore removes dimensions from them.
<img width="252" alt="Screenshot 2022-11-09 at 10 51 18" src="https://user-images.githubusercontent.com/9356806/200916052-49718109-d5cb-4b9b-a8c1-4f6f1493bab5.png">

Reply:
oh, I was saying maybe ``v_list = tree.flatten(v)`` will just work here? do you really need ``[v]`` here?

Reply:
Ah right, changed! Thanks :)